ClusterHealthRestCancellationIT.testClusterHealthRestCancellation failed

### CI Link

https://gradle-enterprise.elastic.co/s/mjyqs6uqr7oam

### Repro line

./gradlew ':qa:smoke-test-http:javaRestTest' --tests "org.elasticsearch.http.ClusterHealthRestCancellationIT.testClusterHealthRestCancellation" -Dtests.seed=E99BACA50D51F65 -Dtests.locale=lt -Dtests.timezone=America/Asuncion -Druntime.java=21

### Does it reproduce?

Yes

### Applicable branches

main

### Failure history

_No response_

### Failure excerpt

```

[2023-09-29T09:17:10,678][INFO ][o.e.g.GatewayService     ] [node_s0] recovered [0] indices into cluster_state
[2023-09-29T09:17:10,701][INFO ][o.e.h.n.s.HealthNodeTaskExecutor] [node_s0] Node [{node_s0}{DgYSsEPtQO2s7lqvvLu7Wg}] is selected as the current health node.
[2023-09-29T09:17:10,761][INFO ][o.e.c.m.MetadataIndexTemplateService] [node_s0] adding template [random_index_template] for index patterns [*]
[2023-09-29T09:17:10,784][INFO ][o.e.h.ClusterHealthRestCancellationIT] [testClusterHealthRestCancellation] [ClusterHealthRestCancellationIT#testClusterHealthRestCancellation]: all set up test
[2023-09-29T09:17:10,794][INFO ][o.e.h.ClusterHealthRestCancellationIT] [testClusterHealthRestCancellation] --> sending cluster state request
[2023-09-29T09:17:10,811][INFO ][o.e.t.TaskAssertions     ] [testClusterHealthRestCancellation] --> waiting for task with prefix [cluster:monitor/health] to start
[2023-09-29T09:17:10,819][INFO ][o.e.h.ClusterHealthRestCancellationIT] [testClusterHealthRestCancellation] --> cancelling cluster health request
[2023-09-29T09:17:10,820][INFO ][o.e.h.ClusterHealthRestCancellationIT] [testClusterHealthRestCancellation] --> checking cluster health task cancelled
[2023-09-29T09:17:10,820][INFO ][o.e.t.TaskAssertions     ] [testClusterHealthRestCancellation] --> checking that all tasks with prefix cluster:monitor/health are marked as cancelled
Rgs 29, 2023 9:17:20 AM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
WARNING: Uncaught exception in thread: Thread[#320,elasticsearch[node_s0][masterService#updateTask][T#1],5,TGRP-ClusterHealthRestCancellationIT]
java.lang.AssertionError: unexpected
	at __randomizedtesting.SeedInfo.seed([E99BACA50D51F65]:0)
	at org.elasticsearch.test.ESTestCase.safeAwait(ESTestCase.java:2012)
	at org.elasticsearch.http.ClusterHealthRestCancellationIT$1.execute(ClusterHealthRestCancellationIT.java:40)
	at org.elasticsearch.cluster.service.MasterService$UnbatchedExecutor.execute(MasterService.java:550)
	at org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:1039)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:1004)
	at org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:232)
	at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.lambda$run$2(MasterService.java:1626)
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:368)
	at org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.run(MasterService.java:1623)
	at org.elasticsearch.cluster.service.MasterService$5.lambda$doRun$0(MasterService.java:1237)
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:368)
	at org.elasticsearch.cluster.service.MasterService$5.doRun(MasterService.java:1216)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CyclicBarrier.dowait(CyclicBarrier.java:259)
	at java.base/java.util.concurrent.CyclicBarrier.await(CyclicBarrier.java:437)
	at org.elasticsearch.test.ESTestCase.safeAwait(ESTestCase.java:2007)
	... 16 more
```

On 8.13 today: https://gradle-enterprise.elastic.co/s/nkfmr7gzg6o6q

```
REPRODUCE WITH: ./gradlew ':qa:smoke-test-http:javaRestTest' --tests "org.elasticsearch.http.ClusterHealthRestCancellationIT.testClusterHealthRestCancellation" -Dtests.seed=9A04368ED6A7E7E -Dtests.locale=fi -Dtests.timezone=Europe/Chisinau -Druntime.java=22

org.elasticsearch.http.ClusterHealthRestCancellationIT > testClusterHealthRestCancellation FAILED
    junit.framework.AssertionFailedError: found no cancellable tasks
        at junit.framework.Assert.fail(Assert.java:57)
        at junit.framework.Assert.assertTrue(Assert.java:22)
        at junit.framework.TestCase.assertTrue(TestCase.java:192)
        at org.elasticsearch.test.TaskAssertions.lambda$assertAllCancellableTasksAreCancelled$2(TaskAssertions.java:82)
        at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1255)
        at org.elasticsearch.test.TaskAssertions.assertAllCancellableTasksAreCancelled(TaskAssertions.java:65)
        at org.elasticsearch.http.ClusterHealthRestCancellationIT.testClusterHealthRestCancellation(ClusterHealthRestCancellationIT.java:85)

    org.elasticsearch.cluster.metadata.ProcessClusterEventTimeoutException: failed to process cluster event (remove-index-template [random_index_template]) within 30s
        at org.elasticsearch.cluster.service.MasterService$TaskTimeoutHandler.doRun(MasterService.java:1460)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.lang.Thread.run(Thread.java:1570)

    java.lang.AssertionError: unexpected
        at __randomizedtesting.SeedInfo.seed([9A04368ED6A7E7E]:0)
        at org.elasticsearch.http.ClusterHealthRestCancellationIT$1.execute(ClusterHealthRestCancellationIT.java:57)

        Caused by:
        java.util.concurrent.TimeoutException
            at java.base/java.util.concurrent.CyclicBarrier.dowait(CyclicBarrier.java:259)
            at java.base/java.util.concurrent.CyclicBarrier.await(CyclicBarrier.java:437)
            at org.elasticsearch.http.ClusterHealthRestCancellationIT$1.execute(ClusterHealthRestCancellationIT.java:52)
```

--- a/qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterHealthRestCancellationIT.java
+++ b/qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/ClusterHealthRestCancellationIT.java
@@ -19,25 +19,15 @@ import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.ClusterStateUpdateTask;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.Priority;
-import org.elasticsearch.test.junit.annotations.TestIssueLogging;
 
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CyclicBarrier;
-import java.util.concurrent.TimeUnit;
 
 import static org.elasticsearch.action.support.ActionTestUtils.wrapAsRestResponseListener;
 import static org.elasticsearch.test.TaskAssertions.assertAllCancellableTasksAreCancelled;
-import static org.elasticsearch.test.TaskAssertions.awaitTaskWithPrefixOnMaster;
 
 public class ClusterHealthRestCancellationIT extends HttpSmokeTestCase {
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/100062")
-    @TestIssueLogging(
-        issueUrl = "https://github.com/elastic/elasticsearch/issues/100062",
-        value = "org.elasticsearch.test.TaskAssertions:TRACE"
-            + ",org.elasticsearch.cluster.service.MasterService:TRACE"
-            + ",org.elasticsearch.tasks.TaskManager:TRACE"
-    )
     public void testClusterHealthRestCancellation() throws Exception {
 
         final var barrier = new CyclicBarrier(2);
@@ -47,18 +37,7 @@ public class ClusterHealthRestCancellationIT extends HttpSmokeTestCase {
                 @Override
                 public ClusterState execute(ClusterState currentState) {
                     safeAwait(barrier);
-                    // safeAwait(barrier);
-
-                    // temporarily lengthen timeout on safeAwait while investigating #100062
-                    try {
-                        barrier.await(60, TimeUnit.SECONDS);
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new AssertionError("unexpected", e);
-                    } catch (Exception e) {
-                        throw new AssertionError("unexpected", e);
-                    }
-
+                    safeAwait(barrier);
                     return currentState;
                 }
 
@@ -72,12 +51,23 @@ public class ClusterHealthRestCancellationIT extends HttpSmokeTestCase {
         clusterHealthRequest.addParameter("wait_for_events", Priority.LANGUID.toString());
 
         final PlainActionFuture<Response> future = new PlainActionFuture<>();
-        logger.info("--> sending cluster state request");
+        logger.info("--> sending cluster health request");
         final Cancellable cancellable = getRestClient().performRequestAsync(clusterHealthRequest, wrapAsRestResponseListener(future));
 
         safeAwait(barrier);
 
-        awaitTaskWithPrefixOnMaster(TransportClusterHealthAction.NAME);
+        // wait until the health request is waiting on the (blocked) master service
+        assertBusy(
+            () -> assertTrue(
+                internalCluster().getCurrentMasterNodeInstance(ClusterService.class)
+                    .getMasterService()
+                    .pendingTasks()
+                    .stream()
+                    .anyMatch(
+                        pendingClusterTask -> pendingClusterTask.source().string().equals("cluster_health (wait_for_events [LANGUID])")
+                    )
+            )
+        );
 
         logger.info("--> cancelling cluster health request");
         cancellable.cancel();

